### PR TITLE
msglist: Offer bulk-mark-as-read button for PM conversations

### PR DESCRIPTION
**Pull request by chrisbobbe**
Fixes: #5156

### CODE REVIEW

**Comment by chrisbobbe** [comment]
Also posted [a thing I noticed while working on this](https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/.60pmNarrowInternal.60's.20jsdoc/near/1295166).

File: src/chat/MarkAsReadButton.js [hunk 1]
```diff
@@ -55,6 +62,23 @@ export default function MarkAsReadButton(props: Props): Node {
...
 
+  const markPmAsRead = useCallback(() => {
+    const userIds = userIdsOfPmNarrow(narrow);
+    let messageIds = undefined;
+    const key = pmUnreadsKeyFromPmKeyIds(userIds, ownUserId);
+    if (userIds.length > 1) {
+      messageIds = unreadHuddles.find(x => x.user_ids_string === key)?.unread_message_ids ?? [];
+    } else {
+      messageIds = unreadPms.find(x => x.sender_id.toString() === key)?.unread_message_ids ?? [];
+    }
```
Comment (gnprice) [inline src/chat/MarkAsReadButton.js:73]: Yikes, this is kind of messy, isn't it. Particularly that gotcha commented on in the first commit.

I'd like to have this kind of code, dealing with the messy structure inside a data model, be located inside that model's code. So e.g. a `getUnreadIdsForPmNarrow`, located among the selectors at the top of `unreadModel.js`.

There's an existing spot where we do something similar, in `getUnreadCountForNarrow` in `unreadSelectors.js`:
```js
      pm: ids => {
        if (ids.length > 1) {
          const unreadsKey = pmUnreadsKeyFromPmKeyIds(ids, ownUserId);
          const unreadItem = unread.huddles.find(x => x.user_ids_string === unreadsKey);
          return unreadItem?.unread_message_ids.length ?? 0;
        } else {
          const senderId = ids[0];
          const unreadItem = unread.pms.find(x => x.sender_id === senderId);
          return unreadItem?.unread_message_ids.length ?? 0;
        }
      },
```

So a good structure would be to factor that logic out of there, and then that case of `getUnreadCountForNarrow` just looks like `ids => getUnreadIdsForPmNarrow(ids).length`.

Note that that also has a variation of how to deal with the string/number thing, that's perhaps a bit cleaner.

File: src/chat/MarkAsReadButton.js [hunk 2]
```diff
@@ -55,6 +62,23 @@ export default function MarkAsReadButton(props: Props): Node {
...
+
+    // The message IDs come from our unread-messages data, which is
+    //   initialized with "only" the most recent 50K unread messages. So
+    //   we'll occasionally, but rarely, miss some messages here; see #5156.
+    // TODO(?): Should we use `queueMarkAsRead` here?
+    api.messagesFlags(auth, messageIds, 'add', 'read');
```
Comment (gnprice) [inline src/chat/MarkAsReadButton.js:79]: Should probably skip this if `messageIds` is empty.

File: src/chat/MarkAsReadButton.js [hunk 3]
```diff
@@ -55,6 +62,23 @@ export default function MarkAsReadButton(props: Props): Node {
...
+    if (userIds.length > 1) {
+      messageIds = unreadHuddles.find(x => x.user_ids_string === key)?.unread_message_ids ?? [];
+    } else {
+      messageIds = unreadPms.find(x => x.sender_id.toString() === key)?.unread_message_ids ?? [];
+    }
+
+    // The message IDs come from our unread-messages data, which is
+    //   initialized with "only" the most recent 50K unread messages. So
+    //   we'll occasionally, but rarely, miss some messages here; see #5156.
+    // TODO(?): Should we use `queueMarkAsRead` here?
```
Comment (gnprice) [inline src/chat/MarkAsReadButton.js:78]: No need for that, I think; the job of that queue is when you're scrolling and have lots of frequent little batches to mark as read.

**Review by gnprice — COMMENTED** [review]
Thanks @chrisbobbe! Comments below.

File: src/chat/MarkAsReadButton.js (hunk 1 above)
Comment (chrisbobbe) [reply src/chat/MarkAsReadButton.js:73]: > Note that that also has a variation of how to deal with the string/number thing, that's perhaps a bit cleaner.

Yeah, maybe. It's not a biggie, but that one doesn't use `pmUnreadsKeyFromPmKeyIds` at all in the 1:1 pm case, even though it's written to be able to handle it.

**Comment by chrisbobbe** [comment]
Thanks! Revision pushed.

**Comment by gnprice** [comment]
Thanks for the revision! Looks good. One nit:
```diff
-  invariant(isPmNarrow(narrow));
```

This gets added with the surrounding code, then removed in the subsequent commit. Did you intend to take it out?

(I think it's fine to either have it or not have it; the next line will throw an error if it's not a PM narrow anyway.)

**Comment by chrisbobbe** [comment]
> Thanks for the revision! Looks good. One nit:
> 
> ```diff
> -  invariant(isPmNarrow(narrow));
> ```
> 
> This gets added with the surrounding code, then removed in the subsequent commit. Did you intend to take it out?
> 
> (I think it's fine to either have it or not have it; the next line will throw an error if it's not a PM narrow anyway.)

Eep, thanks for catching this! Yeah, I think I was doing a complicated reordering of commits, and this fell through. Agreed that we don't need the invariant here.

Small correction: at the commit that includes that invariant line, the function (`getUnreadIdsForPmNarrow`) would actually error unconditionally. I've sent https://github.com/zertosh/invariant/pull/48 for that. 😝 I'm not sure `invariant` is actively maintained; if you'd like, I'll make a little libdef for it with that fix.

**Comment by chrisbobbe** [comment]
Merged, with the invariant line taken out.

**Comment by gnprice** [comment]
> the function (`getUnreadIdsForPmNarrow`) would actually error unconditionally. I've sent [zertosh/invariant#48](https://github.com/zertosh/invariant/pull/48) for that.

Ha, I see.

> I'm not sure `invariant` is actively maintained; if you'd like, I'll make a little libdef for it with that fix.

Could try that. Though I'm not sure a libdef (or that `.js.flow` file) will actually have any effect… because Flow actually special-cases calls to `invariant`, by name:
https://github.com/facebook/flow/blob/d247c5eef542a1745d831edeab4e11e305effaf4/src/typing/statement.ml#L221-L223

So this is likely a bug in that logic in Flow… that or a divergence between the `invariant` NPM package and the Facebook-internal `invariant`. The way to find out which probably runs through filing an issue in Flow.

Related Flow issue, with background on why it's special: https://github.com/facebook/flow/issues/ 112

From back-links in that thread, a possible form of a solution if the NPM package and what's inside Facebook have diverged (other than the solution of making the package align with the Facebook version):
https://github.com/facebook/flow/commit/3e35bd475435a4fd282939be21ebeb3a6fab7106

I haven't really been bothered by the fact that `invariant` is special-cased, because the name is specific enough that it doesn't feel likely to be accidentally used for something unrelated. But if the special-cased behavior is wrong, it'd certainly be good to fix it.

**Pull request merged**